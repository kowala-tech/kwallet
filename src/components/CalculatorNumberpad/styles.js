const styles = (theme) => ({
	root: {
		minWidth: "300px",
		borderTop: "1px solid #000",
		paddingTop: theme.spacing.unit * 2
	},
	keypad: {
	},
	inputKeys: {
	},
	digitKeys: {
		display: "flex",
		flex: "1 auto",
		flexFlow: "row wrap-reverse",
	},
	digitKey: {
	},
	operatorKeys: {
		display: "flex",
		flex: "1 auto",
		flexFlow: "row wrap",
	},
	operatorKey: {
		fontSize: 20
	}
});

export default styles;
