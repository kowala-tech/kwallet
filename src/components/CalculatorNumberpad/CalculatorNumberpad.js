import React from "react";
import PropTypes from "prop-types";
// Material UI Imports
import { withStyles } from "@material-ui/core/styles";
// Component Related Imports
import styles from "./styles";
import CalculatorKey from "./components/CalculatorKey";
import BackSpaceIcon from "@material-ui/icons/Backspace";

const CalculatorOperations = {
	"/": (prevValue, nextValue) => prevValue / nextValue,
	"*": (prevValue, nextValue) => prevValue * nextValue,
	"+": (prevValue, nextValue) => prevValue + nextValue,
	"-": (prevValue, nextValue) => prevValue - nextValue,
	"=": (prevValue, nextValue) => nextValue
};

class CalculatorNumberpad extends React.Component {
	state = {
		value: null,
		displayValue: "0",
		operator: null,
		waitingForOperand: false
	};

	componentDidMount() {
		document.addEventListener("keydown", this.handleKeyDown);
	}

	componentWillUnmount() {
		document.removeEventListener("keydown", this.handleKeyDown);
	}

	clearAll() {
		this.setState({
			value: null,
			displayValue: "0",
			operator: null,
			waitingForOperand: false
		});
		this.props.handleChange("0");
	}

	clearLastChar() {
		const { displayValue } = this.state;
		const newValue = displayValue.substring(0, displayValue.length - 1) || "0";
		this.setState({ displayValue: newValue });
		this.props.handleChange( newValue );
	}

	inputPercent() {
		const { displayValue } = this.state;
		const currentValue = parseFloat(displayValue);

		if (currentValue === 0)
			return;

		const fixedDigits = displayValue.replace(/^-?\d*\.?/, "");
		const parsedValue = parseFloat(displayValue) / 100;
		const newValue = String(parsedValue.toFixed(fixedDigits.length + 2));

		this.setState({ displayValue: newValue });
		this.props.handleChange( newValue );
	}

	inputDot() {
		const { displayValue } = this.state;

		if (!(/\./).test(displayValue)) {
			const newValue = displayValue + ".";
			this.setState({
				displayValue: newValue,
				waitingForOperand: false
			});
			this.props.handleChange( newValue );
		}
	}

	inputDigit(digit) {
		this.props.handleChange(digit);
		const { displayValue, waitingForOperand } = this.state;
		if (waitingForOperand) {
			const newValue = String(digit);
			this.setState({
				displayValue: newValue,
				waitingForOperand: false
			});
			this.props.handleChange( newValue );
		} else {
			const newValue = displayValue === "0" ? String(digit) : displayValue + digit;
			this.setState({ displayValue: newValue });
			this.props.handleChange( newValue );
		}
	}

	performOperation(nextOperator) {
		const { value, displayValue, operator } = this.state;
		const inputValue = parseFloat(displayValue);

		if (value == null) {
			this.setState({
				value: inputValue
			});
		} else if (operator) {
			const currentValue = value || 0;
			const newValue = CalculatorOperations[operator](currentValue, inputValue);
			this.setState({
				value: newValue,
				displayValue: String(newValue)
			});
			this.props.handleChange( newValue );
		}

		this.setState({
			waitingForOperand: true,
			operator: nextOperator
		});
	}

	handleKeyDown = (event) => {
		let { key } = event;

		if (key === "Enter")
			key = "=";

		if ((/\d/).test(key)) {
			event.preventDefault();
			this.inputDigit(parseInt(key, 10));
		} else if (key in CalculatorOperations) {
			event.preventDefault();
			this.performOperation(key);
		} else if (key === ".") {
			event.preventDefault();
			this.inputDot();
		} else if (key === "%") {
			event.preventDefault();
			this.inputPercent();
		} else if (key === "Backspace") {
			event.preventDefault();
			this.clearLastChar();
		} else if (key === "Clear") {
			event.preventDefault();
			this.clearAll();
		}
	};

	render() {
		const { classes } = this.props;

		return (
			<div className={classes.root}>
				<div className={classes.keypad}>
					<div className={classes.inputKeys}>
						<div className={classes.digitKeys}>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDot()}>∙</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(0)}>0</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.clearLastChar()}><BackSpaceIcon /></CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(1)}>1</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(2)}>2</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(3)}>3</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(4)}>4</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(5)}>5</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(6)}>6</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(7)}>7</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(8)}>8</CalculatorKey>
							<CalculatorKey className={classes.digitKey}
								onPress={() => this.inputDigit(9)}>9</CalculatorKey>
						</div>
					</div>
					<div className={classes.operatorKeys}>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("%")}>%</CalculatorKey>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("/")}>÷</CalculatorKey>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("*")}>×</CalculatorKey>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("-")}>−</CalculatorKey>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("+")}>+</CalculatorKey>
						<CalculatorKey className={classes.operatorKey}
							onPress={() => this.performOperation("=")}>=</CalculatorKey>
					</div>
				</div>
			</div>
		);
	}
}

CalculatorNumberpad.propTypes = {
	classes: PropTypes.object.isRequired,
	handleChange: PropTypes.func.isRequired
};

export default withStyles(styles)(CalculatorNumberpad);
