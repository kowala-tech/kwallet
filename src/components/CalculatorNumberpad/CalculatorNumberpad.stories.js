import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
// Story related component imports
import CalculatorNumberpad from "./CalculatorNumberpad";

storiesOf("Calculator Numberpad", module)
	.add("default", () =>
		(<CalculatorNumberpad
			handleChange={action()}
		/>)
	);
