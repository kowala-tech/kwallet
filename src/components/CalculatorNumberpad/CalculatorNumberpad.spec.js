import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
// Story related component imports
import CalculatorNumberpad from "./CalculatorNumberpad";

describe("CalculatorNumberpad", () => {

	it("it should match the snapshot", () => {
		const component = toJson(
			shallow(
				<CalculatorNumberpad
					handleChange={jest.fn()}
				/>
			)
		);
		expect(component).toMatchSnapshot();
	});

});
