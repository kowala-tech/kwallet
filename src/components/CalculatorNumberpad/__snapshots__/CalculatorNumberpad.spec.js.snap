// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorNumberpad it should match the snapshot 1`] = `
<CalculatorNumberpad
  classes={
    Object {
      "digitKey": "CalculatorNumberpad-digitKey-5",
      "digitKeys": "CalculatorNumberpad-digitKeys-4",
      "inputKeys": "CalculatorNumberpad-inputKeys-3",
      "keypad": "CalculatorNumberpad-keypad-2",
      "operatorKey": "CalculatorNumberpad-operatorKey-7",
      "operatorKeys": "CalculatorNumberpad-operatorKeys-6",
      "root": "CalculatorNumberpad-root-1",
    }
  }
  handleChange={[MockFunction]}
/>
`;
